File: src/hackerrank/practice/IceCreamParlor.java
package hackerrank.practice;

import java.util.HashMap;
import java.util.Map;

public class IceCreamParlor {

    public static void main(String[] args) {
        whatFlavors(new int[] {1 ,4 ,5 ,3 ,2}, 4);
    }

    // Complete the whatFlavors function below.
    static void whatFlavors(int[] cost, int money) {

        Map<Integer, Integer> map = new HashMap<>();

        for (int c : cost) {
            map.put(c, map.getOrDefault(c, 0) + 1);
        }

        int idxA =0, idxB =0;

        for (int i = 0;i< cost.length;i++) {

            int key = cost[i];

            // Remove key
            int count = map.get(key);
            if (count == 1) {
                map.remove(key);
            } else {
                map.put(key, count - 1);
            }
            // Check
            if (map.containsKey(money - key)) {
                idxA = i;
                break;
            }

            // Restore key
            map.put(key, map.getOrDefault(key, 0) + 1);
        }

        for (int i = 0;i<cost.length;i++) {
            if (i == idxA) {
                continue;
            }

            if (money == cost[idxA] + cost[i]) {
                idxB = i;
                break;
            }
        }

        System.out.println(String.format("%d %d", idxA + 1,idxB + 1));

    }
}

File: src/hackerrank/practice/IceCreamParlor.kt
package hackerrank.practice

// Complete the icecreamParlor function below.
fun icecreamParlor(m: Int, arr: Array<Int>): Array<Int> {

    val map = arr.groupingBy { it }.eachCount().toMutableMap()

    var costs = mutableListOf<Int>()

    for (a in arr) {
        map[a] = map[a]!! - 1

        if (map.getOrDefault(m - a, 0) != 0) {
            costs.apply {
                add(a)
                add(m -a)
            }
            break
        }
        map[a] = map[a]!! + 1
    }

    val a = getIndex(arr, costs[0], 0)
    val b = getIndex(arr , costs[1], a)

    return arrayOf<Int>(a, b).apply {
        sort()
    }
}

fun getIndex(arr: Array<Int>, c: Int, startIndex : Int): Int {

    for (i in startIndex until  arr.size) {
        if (arr[i] == c) {
            return i + 1
        }
    }
    return -1

}
